README.txt
= ImageScience

home :: https://github.com/seattlerb/image_science
bugs :: https://github.com/seattlerb/image_science/issues
rdoc :: http://docs.seattlerb.org/image_science

== DESCRIPTION:

ImageScience is a clean and happy Ruby library that generates
thumbnails -- and kicks the living crap out of RMagick. Oh, and it
doesn't leak memory like a sieve. :)

For more information including build steps, see http://seattlerb.rubyforge.org/

== FEATURES/PROBLEMS:

* Glorious graphics manipulation magi... errr, SCIENCE! in less than 300 LoC!
* Supports square and proportional thumbnails, as well as arbitrary resizes.
* Pretty much any graphics format you could want. No really.

== SYNOPSIS:

  ImageScience.with_image file do |img|
    img.cropped_thumbnail 100 do |thumb|
      thumb.save "#{file}_cropped.png"
    end

    img.thumbnail 100 do |thumb|
      thumb.save "#{file}_thumb.png"
    end
  end

== REQUIREMENTS:

* FreeImage
* ImageScience

== INSTALL:

* Download and install FreeImage. See notes at url above.
* sudo gem install -y image_science
* see http://seattlerb.rubyforge.org/ImageScience.html for more info.

== LICENSE:

(The MIT License)

Copyright (c) 2006-2009 Ryan Davis, Seattle.rb

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 1.3.2 / 2022-10-01

* 2 bug fixes:

  * Fix include dir finder to work w/ M1 homebrew setup better.
  * Switch from FreeImage_RotateClassic to FreeImage_Rotate. (ivanetchart)

=== 1.3.1 / 2019-12-14

* 1 bug fix:

  * Added noreturn attribute to raise_error since it always calls rb_raise.

=== 1.3.0 / 2016-05-16

* 1 minor enhancement:

  * Add rotate method. (bhenderson)

* 1 bug fix:

  * Fixed error-raising behaviour. (omp)

=== 1.2.6 / 2014-01-15

* 1 bug fix:

  * Defer raising exception until wrapped by rb_ensure. (nocode)

=== 1.2.5 / 2013-06-20

* 1 bug fix:

  * Fixed memory leak in the ReOrient function. (Blargel/abnerqian)

=== 1.2.4 / 2012-11-23

* 1 minor enhancement:

  * Extract out orientation logic and add it to with_image_from_memory. (bhenderson)

=== 1.2.3 / 2012-05-07

* 2 bug fixes:

  * Fixed thumbnail float imprecision errors. (jdelStrother)
  * Switched to using tmpdir for INLINEDIR to avoid failures in CI environments w/o homedirs

=== 1.2.2 / 2012-03-13

* 1 minor enhancement:

  * Extended build to look for homebrew setup (hcatlin)

* 2 bug fixes:

  * Fixed 1.9 warnings, clang warnings, etc...
  * Fixed compilation error when ruby's config mandates C89 style decls. (nocode)

=== 1.2.1 / 2009-08-14

* 2 minor enhancements:

  * Added luis' patches to make it build properly on windows.
  * with_image now raises on missing/bad files.

=== 1.2.0 / 2009-06-23

* 7 minor enhancements:

  * Moved quick_thumb to bin/image_science_thumb and properly added.
  * Added -s (square) flag to bin/image_science_thumb
  * Added autorotating on image load. (choonkeat)
  * Added ruby_inline to clean globs
  * Added with_image_from_memory. (sumbach)
  * Switched to minitest.
  * Updated rakefile for now hoe capabilities.

* 3 bug fixes:

  * Check and convert to 24 BPP if save type is jpg. Caused by 32bpp png to jpg.
  * Fixed 1.9isms
  * Fixed BMP support. Tweaked whitespace.

=== 1.1.3 / 2007-05-30

* 2 minor enhancements:

  * Added quick_thumb as an example to look at.
  * Error handler doesn't raise by default. Raises if $DEBUG==true.

=== 1.1.2 / 2007-04-18

* 2 bug fixes:

  * reports bad height/width values for resize
  * explicitly removes ICC color profiles from PNGs (bug in freeimage).

=== 1.1.1 / 2007-03-08

* 5 minor enhancements:

  * Added error handler that raises with information about what went wrong.
  * thumbnail is now pure ruby, everything now uses resize.
  * Produces cleaner JPEG files, with a small cost to file size/speed.
  * resize now uses Catmull-Rom spline filter for better quality.
  * resize copies existing ICC Profile to thumbnail, producing better color.
  * ICC Profile NOT copied for PNG as it seems to be buggy.

* 1 bug fix:

  * Fixed rdoc

=== 1.1.0 / 2007-01-05

* 3 major enhancements:

  * Added resize(width, height)
  * Added save(path)
  * All thumbnail and resize methods yield instead of saving directly.

* 1 minor enhancement:

  * Will now try to use FreeImage from ports if /opt/local exists.

* 2 bug fixes:

  * Fixed the linker issue on PPC.
  * Rakefile will now clean the image files created by bench.rb

=== 1.0.0 / 2006-12-01

* 1 major enhancement

  * Birthday!



test/test_image_science.rb
require 'minitest'
require 'tmpdir'
dir = Dir.mktmpdir "image_science."
ENV['INLINEDIR'] = dir
Minitest.after_run do
  require 'fileutils'
  FileUtils.rm_rf dir
end

require 'rubygems'
require 'minitest/test'
require 'minitest/autorun' if $0 == __FILE__
require 'image_science'

class TestImageScience < Minitest::Test
  def setup
    @path = 'test/pix.png'
    @tmppath = 'test/pix-tmp.png'
    @tmpjpeg = 'test/pix-tmp.jpg'
    @h = @w = 50
  end

  def teardown
    File.unlink @tmppath if File.exist? @tmppath
    File.unlink @tmpjpeg if File.exist? @tmpjpeg
  end

  def test_class_with_image
    buffer = nil
    ImageScience.with_image @path do |img|
      assert_kind_of ImageScience, img
      assert_equal @h, img.height
      assert_equal @w, img.width
      assert img.save(@tmppath)
    end

    assert File.exist?(@tmppath)

    ImageScience.with_image @tmppath do |img|
      assert_kind_of ImageScience, img
      assert_equal @h, img.height
      assert_equal @w, img.width
    end
  end

  def test_class_with_image_missing
    assert_raises TypeError do
      ImageScience.with_image @path + "nope" do |img|
        flunk
      end
    end
  end

  def test_class_with_image_missing_with_img_extension
    assert_raises RuntimeError do
      assert_nil ImageScience.with_image("nope#{@path}") do |img|
        flunk
      end
    end
  end

  def test_class_with_image_from_memory
    data = File.binread(@path)

    ImageScience.with_image_from_memory data do |img|
      assert_kind_of ImageScience, img
      assert_equal @h, img.height
      assert_equal @w, img.width
      assert img.save(@tmppath)
    end

    assert File.exist?(@tmppath)

    ImageScience.with_image @tmppath do |img|
      assert_kind_of ImageScience, img
      assert_equal @h, img.height
      assert_equal @w, img.width
    end
  end

  def test_class_with_image_from_memory_empty_string
    assert_raises TypeError do
      ImageScience.with_image_from_memory "" do |img|
        flunk
      end
    end
  end

  def test_resize
    ImageScience.with_image @path do |img|
      img.resize(25, 25) do |thumb|
        assert thumb.save(@tmppath)
      end
    end

    assert File.exist?(@tmppath)

    ImageScience.with_image @tmppath do |img|
      assert_kind_of ImageScience, img
      assert_equal 25, img.height
      assert_equal 25, img.width
    end
  end

  def test_resize_floats
    ImageScience.with_image @path do |img|
      img.resize(25.2, 25.7) do |thumb|
        assert thumb.save(@tmppath)
      end
    end

    assert File.exist?(@tmppath)

    ImageScience.with_image @tmppath do |img|
      assert_kind_of ImageScience, img
      assert_equal 25, img.height
      assert_equal 25, img.width
    end
  end

  def test_resize_zero
    assert_raises ArgumentError do
      ImageScience.with_image @path do |img|
        img.resize(0, 25) do |thumb|
          assert thumb.save(@tmppath)
        end
      end
    end

    refute File.exist?(@tmppath)

    assert_raises ArgumentError do
      ImageScience.with_image @path do |img|
        img.resize(25, 0) do |thumb|
          assert thumb.save(@tmppath)
        end
      end
    end

    refute File.exist?(@tmppath)
  end

  def test_resize_negative
    assert_raises ArgumentError do
      ImageScience.with_image @path do |img|
        img.resize(-25, 25) do |thumb|
          assert thumb.save(@tmppath)
        end
      end
    end

    refute File.exist?(@tmppath)

    assert_raises ArgumentError do
      ImageScience.with_image @path do |img|
        img.resize(25, -25) do |thumb|
          assert thumb.save(@tmppath)
        end
      end
    end

    refute File.exist?(@tmppath)
  end

  def test_thumbnail
    ImageScience.with_image @path do |img|
      img.thumbnail(29) do |thumb|
        assert thumb.save(@tmppath)
      end
    end

    assert File.exist?(@tmppath)

    ImageScience.with_image @tmppath do |img|
      assert_kind_of ImageScience, img
      assert_equal 29, img.height
      assert_equal 29, img.width
    end
  end

  def test_auto_rotate_from_file
    ImageScience.with_image "test/portrait.jpg" do |img|
      assert_equal 50, img.height
      assert_equal 38, img.width
    end
  end

  def test_auto_rotate_from_memory
    data = File.binread("test/portrait.jpg")
    ImageScience.with_image_from_memory data do |img|
      assert_equal 50, img.height
      assert_equal 38, img.width
    end
  end

  def test_rotate
    ImageScience.with_image "test/portrait.jpg" do |image|
      image.rotate 90 do |img|
        assert_equal 50, img.width
        assert_equal 38, img.height
      end
    end
  end

  def test_rotate_not_45
    e = assert_raises ArgumentError do
      ImageScience.with_image "test/portrait.jpg" do |image|
        image.rotate 44
      end
    end

    assert_equal 'Angle must be 45 degree skew', e.message
  end

  def test_buffer
    buffer = nil
    ImageScience.with_image @path do |img|
      img.thumbnail(128) do |thumb|
        img.save(@tmpjpeg)
        buffer = img.buffer('.jpg')
      end
    end

    file_data = File.binread(@tmpjpeg)
    assert_equal file_data, buffer
  end

  def test_buffer_default
    buffer = nil
    ImageScience.with_image @path do |img|
      img.thumbnail(128) do |thumb|
        img.save(@tmppath)
        buffer = img.buffer
      end
    end

    file_data = File.binread(@tmppath)
    assert_equal file_data, buffer
  end
end


bench.rb
#!/usr/local/bin/ruby -w

require 'benchmark'
require 'rubygems'
require 'image_science'

max = (ARGV.shift || 100).to_i
ext = ARGV.shift || "png"

file = "blah_big.#{ext}"

if RUBY_PLATFORM =~ /darwin/ then
  # how fucking cool is this???
  puts "taking screenshot for thumbnailing benchmarks"
  system "screencapture -SC #{file}"
else
  abort "You need to plonk down #{file} or buy a mac"
end unless test ?f, "#{file}"

ImageScience.with_image(file.sub(/#{ext}$/, 'png')) do |img|
  img.save(file)
end if ext != "png"

puts "# of iterations = #{max}"
Benchmark::bm(20) do |x|
  x.report("null_time") {
    for i in 0..max do
      # do nothing
    end
  }

  x.report("cropped") {
    for i in 0..max do
      ImageScience.with_image(file) do |img|
        img.cropped_thumbnail(100) do |thumb|
          thumb.save("blah_cropped.#{ext}")
        end
      end
    end
  }

  x.report("proportional") {
    for i in 0..max do
      ImageScience.with_image(file) do |img|
        img.thumbnail(100) do |thumb|
          thumb.save("blah_thumb.#{ext}")
        end
      end
    end
  }

  x.report("resize") {
    for i in 0..max do
      ImageScience.with_image(file) do |img|
        img.resize(200, 200) do |resize|
          resize.save("blah_resize.#{ext}")
        end
      end
    end
  }
end

# File.unlink(*Dir["blah*#{ext}"])


lib/image_science.rb
#!/usr/local/bin/ruby -w

require 'rubygems'
require 'inline'

##
# Provides a clean and simple API to generate thumbnails using
# FreeImage as the underlying mechanism.
#
# For more information or if you have build issues with FreeImage, see
# http://seattlerb.rubyforge.org/ImageScience.html

class ImageScience
  VERSION = '1.3.2'

  ##
  # The top-level image loader opens +path+ and then yields the image.
  #
  # :singleton-method: with_image

  ##
  # The top-level image loader, opens an image from the string +data+
  # and then yields the image.
  #
  # :singleton-method: with_image_from_memory

  ##
  # Crops an image to +left+, +top+, +right+, and +bottom+ and then
  # yields the new image.
  #
  # :method: with_crop

  ##
  # Returns the width of the image, in pixels.
  #
  # :method: width

  ##
  # Returns the height of the image, in pixels.
  #
  # :method: height

  ##
  # Saves the image out to +path+. Changing the file extension will
  # convert the file type to the appropriate format.
  #
  # :method: save

  ##
  # Resizes the image to +width+ and +height+ using a cubic-bspline
  # filter and yields the new image.
  #
  # :method: resize

  ##
  # Rotate the image to +angle+. Limited to 45 degree skewing only.
  #
  # :method: rotate

  ##
  # Creates a proportional thumbnail of the image scaled so its longest
  # edge is resized to +size+ and yields the new image.

  def thumbnail(size) # :yields: image
    w, h = width, height
    scale = size.to_f / (w > h ? w : h)

    self.resize((w * scale).round, (h * scale).round) do |image|
      yield image
    end
  end

  ##
  # Creates a square thumbnail of the image cropping the longest edge
  # to match the shortest edge, resizes to +size+, and yields the new
  # image.

  def cropped_thumbnail(size) # :yields: image
    w, h = width, height
    l, t, r, b, half = 0, 0, w, h, (w - h).abs / 2

    l, r = half, half + h if w > h
    t, b = half, half + w if h > w

    with_crop(l, t, r, b) do |img|
      img.thumbnail(size) do |thumb|
        yield thumb
      end
    end
  end

  inline do |builder|
    %w[/opt/homebrew /opt/local /usr/local].each do |dir|
      if File.directory? "#{dir}/include" then
        builder.add_compile_flags "-I#{dir}/include"
        builder.add_link_flags "-L#{dir}/lib"
      end
    end

    builder.add_link_flags "-lfreeimage"
    unless RUBY_PLATFORM =~ /mswin/
      builder.add_link_flags "-lfreeimage"
      # TODO: detect PPC
      builder.add_link_flags "-lstdc++" # only needed on PPC for some reason
    else
      builder.add_link_flags "freeimage.lib"
    end
    builder.include '"FreeImage.h"'

    builder.prefix <<-"END"
      #define GET_BITMAP(name) Data_Get_Struct(self, FIBITMAP, (name)); if (!(name)) rb_raise(rb_eTypeError, "Bitmap has already been freed");
      static ID err_key; /* used as thread-local key */
      static void clear_error(void);
      static void raise_error(void);
    END

    builder.prefix <<-"END"
      VALUE unload(VALUE self) {
        FIBITMAP *bitmap;
        GET_BITMAP(bitmap);

        FreeImage_Unload(bitmap);
        DATA_PTR(self) = NULL;
        clear_error();
        return Qnil;
      }
    END

    builder.prefix <<-"END"
      VALUE wrap_and_yield(FIBITMAP *image, VALUE self, FREE_IMAGE_FORMAT fif) {
        unsigned int self_is_class = rb_type(self) == T_CLASS;
        VALUE klass = self_is_class ? self         : CLASS_OF(self);
        VALUE type  = self_is_class ? INT2FIX(fif) : rb_iv_get(self, "@file_type");
        VALUE obj = Data_Wrap_Struct(klass, NULL, NULL, image);
        rb_iv_set(obj, "@file_type", type);
        return rb_ensure(rb_yield, obj, unload, obj);
      }
    END

    builder.prefix <<-"END"
      void copy_icc_profile(VALUE self, FIBITMAP *from, FIBITMAP *to) {
        FREE_IMAGE_FORMAT fif = FIX2INT(rb_iv_get(self, "@file_type"));
        if (fif != FIF_PNG && FreeImage_FIFSupportsICCProfiles(fif)) {
          FIICCPROFILE *profile = FreeImage_GetICCProfile(from);
          if (profile && profile->data) {
            FreeImage_CreateICCProfile(to, profile->data, profile->size);
          }
        }
      }
    END

    # we defer raising the error until it we find a safe point to do so
    # We cannot use rb_ensure in these cases because FreeImage may internally
    # make allocations via which our code will never see.
    builder.prefix <<-"END"
      void FreeImageErrorHandler(FREE_IMAGE_FORMAT fif, const char *message) {
        VALUE err = rb_sprintf(
                 "FreeImage exception for type %s: %s",
                  (fif == FIF_UNKNOWN) ? "???" : FreeImage_GetFormatFromFIF(fif),
                  message);
        rb_thread_local_aset(rb_thread_current(), err_key, err);
      }
    END

    # do not call this until necessary variables are wrapped up for GC
    # otherwise there will be leaks
    builder.prefix <<-"END"
      __attribute__((__noreturn__))
      static void raise_error(void) {
        VALUE err = rb_thread_local_aref(rb_thread_current(), err_key);
        if (NIL_P(err)) {
          rb_raise(rb_eRuntimeError, "FreeImage exception");
        } else {
          rb_thread_local_aset(rb_thread_current(), err_key, Qnil);
          rb_raise(rb_eRuntimeError, "%s", StringValueCStr(err));
        }
      }
    END

    builder.prefix <<-"END"
      static void clear_error(void) {
        if (!NIL_P(rb_thread_local_aref(rb_thread_current(), err_key))) {
          rb_thread_local_aset(rb_thread_current(), err_key, Qnil);
        }
      }
    END

    builder.prefix <<-"END"
      FIBITMAP* ReOrient(FIBITMAP *bitmap) {
        FITAG *tagValue = NULL;
        FIBITMAP *oldBitmap = bitmap;
        FreeImage_GetMetadata(FIMD_EXIF_MAIN, bitmap, "Orientation", &tagValue);
        switch (tagValue == NULL ? 0 : *((short *) FreeImage_GetTagValue(tagValue))) {
          case 6:
            bitmap = FreeImage_Rotate(bitmap, 270, NULL);
            break;
          case 3:
            bitmap = FreeImage_Rotate(bitmap, 180, NULL);
            break;
          case 8:
            bitmap = FreeImage_Rotate(bitmap, 90, NULL);
            break;
          default:
            bitmap = FreeImage_Clone(bitmap);
            break;
        }
        FreeImage_Unload(oldBitmap);
        return bitmap;
      }
    END

    builder.add_to_init "FreeImage_SetOutputMessage(FreeImageErrorHandler);"
    builder.add_to_init 'err_key = rb_intern("__FREE_IMAGE_ERROR");'

    builder.c_singleton <<-"END"
      VALUE with_image(char * input) {
        FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;
        int flags;

        fif = FreeImage_GetFileType(input, 0);
        if (fif == FIF_UNKNOWN) fif = FreeImage_GetFIFFromFilename(input);
        if ((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsReading(fif)) {
          FIBITMAP *bitmap;
          VALUE result = Qnil;
          flags = fif == FIF_JPEG ? JPEG_ACCURATE : 0;

          if (!(bitmap = FreeImage_Load(fif, input, flags))) raise_error();
          if (!(bitmap = ReOrient(bitmap))) raise_error();

          result = wrap_and_yield(bitmap, self, fif);
          return result;
        }
        rb_raise(rb_eTypeError, "Unknown file format");
        return Qnil;
      }
    END

    builder.c_singleton <<-"END"
      VALUE with_image_from_memory(VALUE image_data) {
        FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;
        BYTE *image_data_ptr;
        DWORD image_data_length;
        FIMEMORY *stream;
        FIBITMAP *bitmap = NULL;
        VALUE result = Qnil;
        int flags;

        Check_Type(image_data, T_STRING);
        image_data_ptr    = (BYTE*)RSTRING_PTR(image_data);
        image_data_length = (DWORD)RSTRING_LEN(image_data);
        stream = FreeImage_OpenMemory(image_data_ptr, image_data_length);

        if (NULL == stream) {
          rb_raise(rb_eTypeError, "Unable to open image_data");
        }

        fif = FreeImage_GetFileTypeFromMemory(stream, 0);
        if ((fif == FIF_UNKNOWN) || !FreeImage_FIFSupportsReading(fif)) {
          FreeImage_CloseMemory(stream);
          rb_raise(rb_eTypeError, "Unknown file format");
        }

        flags = fif == FIF_JPEG ? JPEG_ACCURATE : 0;
        bitmap = FreeImage_LoadFromMemory(fif, stream, flags);
        FreeImage_CloseMemory(stream);

        if (!bitmap) raise_error();
        if (!(bitmap = ReOrient(bitmap))) raise_error();

        result = wrap_and_yield(bitmap, self, fif);
        return result;
      }
    END

    builder.c <<-"END"
      VALUE with_crop(int l, int t, int r, int b) {
        FIBITMAP *copy, *bitmap;
        GET_BITMAP(bitmap);

        if (!(copy = FreeImage_Copy(bitmap, l, t, r, b))) raise_error();

        copy_icc_profile(self, bitmap, copy);
        return wrap_and_yield(copy, self, 0);
      }
    END

    builder.c <<-"END"
      int height() {
        FIBITMAP *bitmap;
        GET_BITMAP(bitmap);

        return FreeImage_GetHeight(bitmap);
      }
    END

    builder.c <<-"END"
      int width() {
        FIBITMAP *bitmap;
        GET_BITMAP(bitmap);

        return FreeImage_GetWidth(bitmap);
      }
    END

    builder.c <<-"END"
      VALUE resize(int w, int h) {
        FIBITMAP *bitmap, *image;
        if (w <= 0) rb_raise(rb_eArgError, "Width <= 0");
        if (h <= 0) rb_raise(rb_eArgError, "Height <= 0");
        GET_BITMAP(bitmap);

        image = FreeImage_Rescale(bitmap, w, h, FILTER_CATMULLROM);
        if (!image) raise_error();

        copy_icc_profile(self, bitmap, image);
        return wrap_and_yield(image, self, 0);
      }
    END

    builder.c <<-"END"
      VALUE rotate(int angle) {
        FIBITMAP *bitmap, *image;
        if ((angle % 45) != 0) rb_raise(rb_eArgError, "Angle must be 45 degree skew");
        GET_BITMAP(bitmap);
        image = FreeImage_Rotate(bitmap, angle, NULL);
        if (image) {
          copy_icc_profile(self, bitmap, image);
          return wrap_and_yield(image, self, 0);
        }
        return Qnil;
      }
    END

    builder.c <<-"END"
      VALUE save(char * output) {
        int flags;
        FIBITMAP *bitmap;
        FREE_IMAGE_FORMAT fif = FreeImage_GetFIFFromFilename(output);
        if (fif == FIF_UNKNOWN) fif = FIX2INT(rb_iv_get(self, "@file_type"));
        if ((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsWriting(fif)) {
          BOOL result = 0, unload = 0;
          GET_BITMAP(bitmap);
          flags = fif == FIF_JPEG ? JPEG_QUALITYSUPERB : 0;

          if (fif == FIF_PNG) FreeImage_DestroyICCProfile(bitmap);
          if (fif == FIF_JPEG && FreeImage_GetBPP(bitmap) != 24) {
            bitmap = FreeImage_ConvertTo24Bits(bitmap), unload = 1; // sue me
            if (!bitmap) raise_error();
          }

          result = FreeImage_Save(fif, bitmap, output, flags);
          if (unload) FreeImage_Unload(bitmap);
          if (!result) raise_error();

          return Qtrue;
        }
        rb_raise(rb_eTypeError, "Unknown file format");
        return Qnil;
      }
    END
  end
end


ext/image_science/image_science.c
#include "ruby.h"
#include "FreeImage.h"

#define GET_BITMAP(name) Data_Get_Struct(self, FIBITMAP, (name)); if (!(name)) rb_raise(rb_eTypeError, "Bitmap has already been freed");

VALUE unload(VALUE self) {
  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);

  FreeImage_Unload(bitmap);
  DATA_PTR(self) = NULL;
  return Qnil;
}


VALUE wrap_and_yield(FIBITMAP *image, VALUE self, FREE_IMAGE_FORMAT fif) {
  unsigned int self_is_class = rb_type(self) == T_CLASS;
  VALUE klass = self_is_class ? self         : CLASS_OF(self);
  VALUE type  = self_is_class ? INT2FIX(fif) : rb_iv_get(self, "@file_type");
  VALUE obj = Data_Wrap_Struct(klass, NULL, NULL, image);
  rb_iv_set(obj, "@file_type", type);
  return rb_ensure(rb_yield, obj, unload, obj);
}


void copy_icc_profile(VALUE self, FIBITMAP *from, FIBITMAP *to) {
  FREE_IMAGE_FORMAT fif = FIX2INT(rb_iv_get(self, "@file_type"));
  if (fif != FIF_PNG && FreeImage_FIFSupportsICCProfiles(fif)) {
    FIICCPROFILE *profile = FreeImage_GetICCProfile(from);
    if (profile && profile->data) {
      FreeImage_CreateICCProfile(to, profile->data, profile->size);
    }
  }
}

static char* lastFreeImageMessage = NULL;

void FreeImageMessageHandler(FREE_IMAGE_FORMAT fif, const char *message) {
  if (lastFreeImageMessage != NULL)
    free(lastFreeImageMessage);

  lastFreeImageMessage = strdup(message);
}


static VALUE with_image(VALUE self, VALUE _input) {
  char * input = StringValuePtr(_input);

  FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;
  int flags;

  fif = FreeImage_GetFileType(input, 0);
  if (fif == FIF_UNKNOWN) fif = FreeImage_GetFIFFromFilename(input);
  if ((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsReading(fif)) {
    FIBITMAP *bitmap;
    VALUE result = Qnil;
    flags = fif == FIF_JPEG ? JPEG_ACCURATE : 0;
    if (bitmap = FreeImage_Load(fif, input, flags)) {
      FITAG *tagValue = NULL;
      FreeImage_GetMetadata(FIMD_EXIF_MAIN, bitmap, "Orientation", &tagValue);
      switch (tagValue == NULL ? 0 : *((short *) FreeImage_GetTagValue(tagValue))) {
        case 6:
          bitmap = FreeImage_Rotate(bitmap, 270, NULL);
          break;
        case 3:
          bitmap = FreeImage_Rotate(bitmap, 180, NULL);
          break;
        case 8:
          bitmap = FreeImage_Rotate(bitmap, 90, NULL);
          break;
        default:
         break;
      }

      result = wrap_and_yield(bitmap, self, fif);
    } else {
      rb_raise(rb_eTypeError,
              "Failed to load image from file %s: %s",
              input,
              lastFreeImageMessage != NULL ? lastFreeImageMessage : "Unspecified error");
    }

    return (result);
  }
  rb_raise(rb_eTypeError, "Unknown file format");
}


static VALUE with_image_from_memory(VALUE self, VALUE _image_data) {
  VALUE image_data = (_image_data);

  FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;

  Check_Type(image_data, T_STRING);
  BYTE *image_data_ptr    = (BYTE*)RSTRING_PTR(image_data);
  DWORD image_data_length = RSTRING_LEN(image_data);
  FIMEMORY *stream = FreeImage_OpenMemory(image_data_ptr, image_data_length);

  if (NULL == stream) {
    rb_raise(rb_eTypeError, "Unable to open image_data");
  }

  fif = FreeImage_GetFileTypeFromMemory(stream, 0);
  if ((fif == FIF_UNKNOWN) || !FreeImage_FIFSupportsReading(fif)) {
    rb_raise(rb_eTypeError, "Unknown file format");
  }

  FIBITMAP *bitmap = NULL;
  VALUE result = Qnil;
  int flags = fif == FIF_JPEG ? JPEG_ACCURATE : 0;
  bitmap = FreeImage_LoadFromMemory(fif, stream, flags);
  FreeImage_CloseMemory(stream);
  if (bitmap) {
    result = wrap_and_yield(bitmap, self, fif);
  }
  return (result);
}


static VALUE with_crop(VALUE self, VALUE _l, VALUE _t, VALUE _r, VALUE _b) {
  int l = FIX2INT(_l);
  int t = FIX2INT(_t);
  int r = FIX2INT(_r);
  int b = FIX2INT(_b);

  FIBITMAP *copy, *bitmap;
  VALUE result = Qnil;
  GET_BITMAP(bitmap);

  if (copy = FreeImage_Copy(bitmap, l, t, r, b)) {
    copy_icc_profile(self, bitmap, copy);
    result = wrap_and_yield(copy, self, 0);
  }
  return (result);
}


static VALUE height(VALUE self) {

  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);

  return INT2FIX(FreeImage_GetHeight(bitmap));
}


static VALUE width(VALUE self) {

  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);

  return INT2FIX(FreeImage_GetWidth(bitmap));
}

static VALUE resize(VALUE self, VALUE _w, VALUE _h, VALUE greyscale) {
  long w = NUM2LONG(_w);
  long h = NUM2LONG(_h);

  FIBITMAP *bitmap, *image;
  if (w <= 0) rb_raise(rb_eArgError, "Width <= 0");
  if (h <= 0) rb_raise(rb_eArgError, "Height <= 0");
  GET_BITMAP(bitmap);
  image = FreeImage_Rescale(bitmap, w, h, FILTER_CATMULLROM);
  if (image) {
    if (greyscale > 0) {
      RGBQUAD a_colors[64];
      RGBQUAD b_colors[64];

      int grey_point = 192;

      for (int i=grey_point;i<256;i++)
      {
          a_colors[i - grey_point].rgbRed = i;
          a_colors[i - grey_point].rgbGreen = i;
          a_colors[i - grey_point].rgbBlue = i;
          b_colors[i - grey_point].rgbRed = grey_point;
          b_colors[i - grey_point].rgbGreen = grey_point;
          b_colors[i - grey_point].rgbBlue = grey_point;
      }

      FIBITMAP *grey = FreeImage_ConvertToGreyscale(image);
      FreeImage_Unload(image);

      if (grey) {
        int result = FreeImage_ApplyColorMapping(grey, a_colors, b_colors, 64, TRUE, FALSE);
        image = grey;
      }
    }
    copy_icc_profile(self, bitmap, image);
    return (wrap_and_yield(image, self, 0));
  }
  return (Qnil);
}

static VALUE _resize(VALUE self, VALUE _w, VALUE _h) {
  return resize(self, _w, _h, 0);
}

static VALUE save(VALUE self, VALUE _output) {
  char * output = StringValuePtr(_output);

  int flags;
  FIBITMAP *bitmap;
  FREE_IMAGE_FORMAT fif = FreeImage_GetFIFFromFilename(output);
  if (fif == FIF_UNKNOWN) fif = FIX2INT(rb_iv_get(self, "@file_type"));
  if ((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsWriting(fif)) {
    GET_BITMAP(bitmap);
    flags = fif == FIF_JPEG ? JPEG_QUALITYGOOD : 0;
    BOOL result = 0, unload = 0;

    if (fif == FIF_PNG) FreeImage_DestroyICCProfile(bitmap);
    if (fif == FIF_JPEG && FreeImage_GetBPP(bitmap) != 24)
      bitmap = FreeImage_ConvertTo24Bits(bitmap), unload = 1;

    result = FreeImage_Save(fif, bitmap, output, flags);

    if (unload) FreeImage_Unload(bitmap);

    return (result ? Qtrue : Qfalse);
  }
  rb_raise(rb_eTypeError, "Unknown file format");
}



#ifdef __cplusplus
extern "C" {
#endif
  void Init_extension() {
    VALUE c = rb_cObject;
    c = rb_const_get(c, rb_intern("ImageScience"));

    rb_define_method(c, "height", (VALUE(*)(ANYARGS))height, 0);
    rb_define_method(c, "resize", (VALUE(*)(ANYARGS))_resize, 2);
    rb_define_method(c, "resize", (VALUE(*)(ANYARGS))resize, 3);
    rb_define_method(c, "save", (VALUE(*)(ANYARGS))save, 1);
    rb_define_method(c, "width", (VALUE(*)(ANYARGS))width, 0);
    rb_define_method(c, "with_crop", (VALUE(*)(ANYARGS))with_crop, 4);
    rb_define_singleton_method(c, "with_image", (VALUE(*)(ANYARGS))with_image, 1);
    rb_define_singleton_method(c, "with_image_from_memory", (VALUE(*)(ANYARGS))with_image_from_memory, 1);
    FreeImage_SetOutputMessage(FreeImageMessageHandler);

  }
#ifdef __cplusplus
}
#endif
